Refactor unicode.py script

pawroman opened the pull request:
Hi, I noticed that the `unicode.py` script used some deprecated escapes in regular expressions. E.g. `\d`, `\w`, `\.` will be illegal in the future without "raw strings". This is now fixed. I have also cleaned up the script quite a bit.

## Escape deprecation

OK (note the `r`):
`re.compile(r"\d")`

Deprecated (from Python 3.6 onwards, see [here][link1] and [here][link2]):
`re.compile("\d")`.

[link1]: https://docs.python.org/3.6/whatsnew/3.6.html#deprecated-python-behavior
[link2]: https://bugs.python.org/issue27364

This was evident running the script using Python 3.7 like so:

```
$ python3 -Wall unicode.py 
unicode.py:227: DeprecationWarning: invalid escape sequence \w
  re1 = re.compile("^ *([0-9A-F]+) *; *(\w+)")
unicode.py:228: DeprecationWarning: invalid escape sequence \.
  re2 = re.compile("^ *([0-9A-F]+)\.\.([0-9A-F]+) *; *(\w+)")
unicode.py:453: DeprecationWarning: invalid escape sequence \d
  pattern = "for Version (\d+)\.(\d+)\.(\d+) of the Unicode"
```

The documentation states that
> A backslash-character pair that is not a valid escape sequence now generates a DeprecationWarning. Although this will eventually become a SyntaxError, that will not be for several Python releases. 

## Testing

To test my changes, I had to add support for choosing the Unicode version to use. The script will default to latest release (which is 12.0.0 at the moment, repo has 11.0.0 checked in).

The script generates the exact same output for version 11.0.0 with Python 2.7 and 3.7 and no longer generates any deprecation warnings:

```
$ python3 -Wall unicode.py -v 11.0.0
Using Unicode version: 11.0.0
Regenerated tables.rs.
$ git diff tables.rs
$ python2 -Wall unicode.py -v 11.0.0
Using Unicode version: 11.0.0
Regenerated tables.rs.
$ git diff tables.rs
$ python2 --version
Python 2.7.16
$ python3 --version
Python 3.7.3
```

## Extra functionality

Furthermore, the script will check and download the latest Unicode version by default (without the `-v` argument). The `--help` is below:

```
$ ./unicode.py --help
usage: unicode.py [-h] [-v VERSION]

Regenerate Unicode tables (tables.rs).

optional arguments:
  -h, --help            show this help message and exit
  -v VERSION, --version VERSION
                        Unicode version to use (if not specified, defaults to
                        latest available final release).
```

## Cleanups

I have cleaned up the code quite a bit, with Python best practices and code style in mind. I'm happy to provide more details and rationale for all my changes if the reviewers so desire.

One externally visible change is that the Unicode data will now be downloaded into `src/libcore/unicode/downloaded` directory suffixed by Unicode version:

```
$ pwd
.../rust/src/libcore/unicode
$ exa -T downloaded/
downloaded
├── 11.0.0
│  ├── DerivedCoreProperties.txt
│  ├── DerivedNormalizationProps.txt
│  ├── PropList.txt
│  ├── ReadMe.txt
│  ├── Scripts.txt
│  ├── SpecialCasing.txt
│  └── UnicodeData.txt
└── 12.0.0
   ├── DerivedCoreProperties.txt
   ├── DerivedNormalizationProps.txt
   ├── PropList.txt
   ├── ReadMe.txt
   ├── Scripts.txt
   ├── SpecialCasing.txt
   └── UnicodeData.txt
```

rust-highfive commented:
r? @KodrAus

(rust_highfive has picked a reviewer for you, use r? to override)

pawroman commented on src/libcore/unicode/unicode.py:
```diff
@@ -414,74 +550,102 @@ def emit_conversions_module(f, to_upper, to_lower, to_title):
+        derived = load_properties(get_path(UnicodeFiles.DERIVED_CORE_PROPERTIES), want_derived)
+
+        # TODO scripts not used?
+        scripts = load_properties(get_path(UnicodeFiles.SCRIPTS), [])
```
This should probably be addressed by the original authors.

pawroman replied:
I have decided to remove this, as it can be brought back easily if needed.

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/194032846) ([raw log](https://api.travis-ci.com/v3/job/194032846/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:064f0514:start=1555598252801053942,finish=1555598254934744840,duration=2133690898
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
$ export GCP_CACHE_BUCKET=rust-lang-ci-cache
$ export AWS_ACCESS_KEY_ID=AKIA46X5W6CZEJZ6XT55
---

[00:03:43] travis_fold:start:tidy
travis_time:start:tidy
tidy check
[00:03:43] tidy error: /checkout/src/libcore/unicode/unicode.py:463: line longer than 100 chars
[00:03:43] tidy error: /checkout/src/libcore/unicode/unicode.py:629: TODO is deprecated; use FIXME
[00:03:45] some tidy checks failed
[00:03:45] 
[00:03:45] 
[00:03:45] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:03:45] 
[00:03:45] 
[00:03:45] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:03:45] Build completed unsuccessfully in 0:00:45
[00:03:45] Build completed unsuccessfully in 0:00:45
[00:03:45] make: *** [tidy] Error 1
[00:03:45] Makefile:67: recipe for target 'tidy' failed
The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:31ae9ce0
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
Thu Apr 18 14:41:31 UTC 2019
---
travis_time:end:00af8fc0:start=1555598492378398530,finish=1555598492383061211,duration=4662681
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:0f3c2b5c
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:159c54d0
travis_time:start:159c54d0
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:1a9a1a97
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

fbstj commented:
should each of your functions have some doc-strings?

pawroman commented:
> should each of your functions have some doc-strings?

Possibly, I can annotate the most obvious ones but I would have to ask the authors about some of the Unicode table calculations -- to explain what is being done and why.

pawroman commented:
Following the suggestion above, I have added docstrings and type annotations for all functions. To fully comprehend what's going on I had to give the code a few good, thorough reads. This spawned more cleanups and the entire script is now refactored.

fbstj commented on src/libcore/unicode/unicode.py:
```diff
@@ -1,93 +1,286 @@
 
-preamble = '''
+# Note that the "curl" program is required for operation.
+# This script is compatible with Python 2.7 and 3.x.
```
very minor nit: is there a reason these comments aren't part of the doc comment?

pawroman replied:
The doc comment serves dual purpose: it describes the script's purpose and serves as the `argparse` program description (which one can see when running the script with `--help` flag). 

Compatibility notice wouldn't be very helpful once the program runs, and it would probably not even show up if incompatible Python version was used. That being said, the changes I made were meant to support current and future Python 3 versions alike, and the script has no external dependencies outside of the standard library.

Similarly, for curl, the script would just crash if curl wasn't installed. In the example below, I have modified `curl` to read `córl` and the following traceback was printed:

```
Fetching: ftp://ftp.unicode.org/Public/UNIDATA/ReadMe.txt
Traceback (most recent call last):
  File "...rust/src/libcore/unicode/unicode.py", line 868, in <module>
    main()
  File "...rust/src/libcore/unicode/unicode.py", line 812, in main
    unicode_version = fetch_files(args.version)
  File "...rust/src/libcore/unicode/unicode.py", line 147, in fetch_files
    readme_content = subprocess.check_output(("córl", readme_url))
  File "/usr/lib/python3.7/subprocess.py", line 395, in check_output
    **kwargs).stdout
  File "/usr/lib/python3.7/subprocess.py", line 472, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/usr/lib/python3.7/subprocess.py", line 775, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.7/subprocess.py", line 1522, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'córl': 'córl'
```

Whether curl should be replaced by Python's [ftplib](https://docs.python.org/3.7/library/ftplib.html) for maximum compatibility and no external dependencies is debatable -- `curl` is listed in Rust build dependencies.

So, given the above, I figured that a mere comment with a compatibility note would suffice, and its contents did not feel like they belonged to the docstring.

fbstj replied:
thanks for your explanation!

KodrAus commented:
I don't think I've got the python experience to be able to properly review this so I'll assign somebody else based on the commit history...

r? @SimonSapin

SimonSapin commented:
Sorry, I don’t have bandwidth available at the moment for non-trivial Rust reviews.

SimonSapin commented:
r? @varkor, based on git-shortlog

varkor commented:
I'll try to take a look soon.

SimonSapin commented:
> The one thing I might be tempted to do is rewrite this (and the other table-generating scripts) in Rust

Such an effort might be able to rely on (parts of?) [ucd-generate](https://github.com/BurntSushi/ucd-generate) or [unic-ucd](https://github.com/open-i18n/rust-unic/tree/master/unic/ucd).

jonas-schievink commented:
Ping from triage @pawroman, there are still outstanding review items that need to be addressed

pawroman commented:
Thanks for being patient - I couldn't find the time to look into this recently.

@varkor Thanks for reviewing - I will address the remarks soon.

I am tempted to rewrite this in Rust myself, however I wanted to clean up the Python code first (so I understand it better) and then perhaps work on a Rust re-implementation.

pawroman commented on src/libcore/unicode/unicode.py:
```diff
@@ -1,93 +1,286 @@
+
+
+UnicodeFiles.ALL_FILES = tuple(
+    getattr(UnicodeFiles, name) for name in dir(UnicodeFiles)
```
That is true, `dir` basically lists entries in the underlying attribute dictionary of an object. That's only guaranteed to be in order of declaration with Python 3.6 and above. However, I'm not sure if we really need to maintain the ordering, as we only ever use this to iterate over the filenames in no particular order (for version checks and for downloads).

varkor replied:
If ordering is the only thing that might change, that's not an issue — as long as the entries themselves don't change.

pawroman replied:
The tuple created here filters out anything with `_` which is (by convention) anything internal or "nonpublic". Even if the output of `dir` would change, it should not add or omit any "public" (aka. non-`_` prefixed fields).

To be 100% sure, however, I will change `dir` to `__dict__` as the "non-private" contents of that changing would break the very basics of Python semantics... So might be a safer bet. I will also add an `assert` to check the number of items in this tuple.

pawroman commented on src/libcore/unicode/unicode.py:
```diff
@@ -1,93 +1,286 @@
+         querying the unicode.org service).
+    """
+    have_version = check_stored_version(version)
+    if have_version:
```
It's usually considered to be idiomatic Python to avoid this kind of checks (e.g. any `False`-y value means that we don't have any version). In this case, I'd argue `is not None` check is not necessary.

Additionally, the types are actually checked if one has a modern Python editor with typing support. As per `check_stored_version` type signature and docstring, the only falsy value returned can be `None`.

This, of course, naively assumes that whoever edits the file will have an editor with typing capability. And because Rust has no CI for Python that I know of, there's no way of enforcing this contract other than idiomatic Python code style, common sense and dynamic typing support in code editors.

But that's just how dynamically-typed languages work. I'm fine with making it more explicit if you still think it's necessary though :)

varkor replied:
It's been a while since I've written Python, so I'm happy leaving this implicit if that's the norm.

varkor commented:
@pawroman: if you could squash the last two commits (so the submodules never change), I'm happy to approve with the new changes 👍

pawroman commented:
@varkor great, thanks. Submodules changing is no more. I have also bullet-proofed `ALL_UNICODE_FILES` to avoid using `dir`.

varkor commented:
@bors r+

pawroman commented:
Is @bors stuck? It's been a while now...

lzutao commented:
cc @Centril

Centril commented:
@bors r=varkor

bors commented:
:pushpin: Commit 2b47a085dd418447f1dd79986df94dd051f27c79 has been approved by `varkor`

<!-- @bors r=varkor 2b47a085dd418447f1dd79986df94dd051f27c79 -->
<!-- homu: {"type":"Approved","sha":"2b47a085dd418447f1dd79986df94dd051f27c79","approver":"varkor"} -->

bors commented:
:hourglass: Testing commit 2b47a085dd418447f1dd79986df94dd051f27c79 with merge 5a94ebbc9dc43dfa82b165b2c41a115b0e869804...
<!-- homu: {"type":"BuildStarted","head_sha":"2b47a085dd418447f1dd79986df94dd051f27c79","merge_sha":"5a94ebbc9dc43dfa82b165b2c41a115b0e869804"} -->

Centril commented:
@bors retry

bors commented:
:hourglass: Testing commit 2b47a085dd418447f1dd79986df94dd051f27c79 with merge b0bd5f236d9bea38b8c9048f379fec179b09984c...
<!-- homu: {"type":"BuildStarted","head_sha":"2b47a085dd418447f1dd79986df94dd051f27c79","merge_sha":"b0bd5f236d9bea38b8c9048f379fec179b09984c"} -->

pietroalbini commented:
@bors r-

pietroalbini commented:
@bors retry
